### ISSUE TITLE
win32gui.SetForegroundWindow fails if there is a foregroung window from another process

### ISSUE THREAD

**Issue by moden-py** [pywinauto/pywinauto#117]
SWAPY's tests have unexpectedly started to fail, although previously worked fine. An investigation has shown the cause is a system right-side window about the network connection has changed. See the attachment.
![scr1447017151 76](https://cloud.githubusercontent.com/assets/13960701/11031102/d743a8a8-86da-11e5-91cd-1542dd2cc2c8.jpg)

In 3132da99190dac1b9206e6a434fcd91a64a901a2 `win32functions.SetForegroundWindow` was changed to `win32gui.SetForegroundWindow` in `SetFocus` method, due `TypeKeys` worked incorrectly in the next code:
``` python
app = Application().Start(cmd_line=u'"C:\\Windows\\system32\\notepad.exe" ')

app2 = Application().Connect(title=u'new 1 - Notepad++', class_name='Notepad++')
notepad2 = app2[u'Notepad++']
scintilla = notepad2[u'1']
scintilla.TypeKeys('habrahabr2', with_spaces=True)

app.Kill_()
```
With `win32functions.SetForegroundWindow` the text was typed in the first app, not in notepad2.
The switching to `win32gui.SetForegroundWindow` fixes 'type text' problem.

I was manage to reproduce the problem locally, here the steps:
* Start a script with postponed `SetFocus` call.

 ``` python
app = Application().Start(cmd_line=u'"apps\\MFC_samples\\x64\\RowList.exe" ')
time.sleep(3)
app['RowList Sample Application'].SetFocus()
 ```
* Manually open calc.

In result:
```
Traceback (most recent call last):
  File "F:/MoDen/swapy-git/debug.py", line 14, in <module>
    app['RowList Sample Application'].SetFocus()
  File "F:\MoDen\pywinauto\pywinauto\controls\HwndWrapper.py", line 1579, in SetFocus
    win32gui.SetForegroundWindow(self.handle)
pywintypes.error: (2, 'SetForegroundWindow', '\xcd\xe5 \xf3\xe4\xe0\xe5\xf2\xf1\xff \xed\xe0\xe9\xf2\xe8 \xf3\xea\xe0\xe7\xe0\xed\xed\xfb\xe9 \xf4\xe0\xe9\xeb.')
```
Note, running calc with `Application().Start` from the script does not cause the problem.

**Comment by airelil** [comment]
Interesting that I also have cryptic problems with TypeKeys. The test_taskbar tests still could be flaky when I run them on a VM and I noticed that there is a problem with sending an "{ENTER}". Sometimes the "{ENTER}" isn't sent to the main explorer window but to a child window that pops up.

**Comment by moden-py** [comment]
@airelil, I've applied an optimization to the fix. I have seen the problem you describe, when `SetForegroundWindow` does not affect. It happens if call "Detach" different threads immediately after the `SetForegroundWindow`.
I figured out we should wait for `SetForegroundWindow` completed and then detach the threads.
I've added WaitFor with `GetForegroundWindow` to check a foreground window has changed.
If you mean `test_taskbar.TaskbarTestCases.testClock` it is exactly the case. I see my the optimization helps.
Unfortunately, "wait timeout" errors happens on the waiting sometimes. I've just increased the timeout to 1 sec. Will check on AppVeyor.

**Comment by airelil** [comment]
I wonder if win32gui.SetForegroundWindow returns a status like the original MSDN function

**Comment by airelil** [comment]
This is from pywin32:
```
  3084 
  3085 // @pyswig HWND|SetForegroundWindow|
  3086 // @pyparm int|hwnd||The handle to the window
  3087 BOOLAPI SetForegroundWindow(HWND hWnd);
  3088 
```
http://pywin32.hg.sourceforge.net/hgweb/pywin32/pywin32/file/055bbedd423c/win32/src/win32gui.i

**Comment by moden-py** [comment]
`win32gui.SetForegroundWindow` returns None if successful and raises an exception if error.

Handled some cases to fix the "wait timeout" errors described behind, also increased `setfocus_timeout` to 2 sec.

**Comment by onuh** [comment]
Pip install keyboard.
Before you set foreground window, simulate a keyboard to esc that is keyboard.send('esc')
You may want to do it three times for either of the following:
1. Sidebar
2. Windows key overlay
3. Task manager which is always on top

**Comment by vasily-v-ryabov** [comment]
Hi @onuh this message is absolutely irrelevant to the issue which was resolved a long time ago. We consider other open-source libraries as our friends (even if they have the same functionality), so we're for fair competition. But this message really looks like a spam because it was duplicated in another old irrelevant issue. Hope it was some mistake or misread. I've deleted the second same message. Thanks for understanding.
